algorithms/ukrainian.sbl

```diff
@@ -0,0 +1,371 @@
+      '{i}{t}{soft}'
+      '{i`}{t}{e}' '{i`}{m}{o}'
+      '{m}{e}'
+      '{l}{a}' '{l}{o}' '{l}{y}' '{l}{u}'
+      '{n}{o}'
+      //'{t}{y}' '{t}{soft}' '{t}{e}' // need to remove ?
+      '{ye}{sh}' '{ye}{m}{o}' '{ye}{t}{e}' '{iu}{t}{soft}'
+      '{e}{sh}' '{e}{m}{o}' '{e}{t}{e}' '{u}{t}{soft}'
+      '{l}{iu}' '{y}{sh}' '{y}{t}{soft}' '{y}{m}{o}' '{y}{t}{e}' '{l}{ia}{t}{soft}' '{sh}{y}'
+      '{l}{ia}{t}{y}'
+      '{yi}{sh}' '{yi}{t}{soft}' '{yi}{m}{o}' '{yi}{t}{e}' '{ia}{t}{soft}' '{ia}{t}{y}'
+        (delete)
+    )
+  )
+
+  define noun as (
+    [substring] among (
+      '{a}{m}' '{a}{m}{y}' '{a}{kh}' // {v}{o}{d}|{a}{m}
+      '{a}{r}' '{a}{r}{e}{m}' '{a}{r}{ia}'
+      '{e}{iu}'
+      '{ia}{m}' '{ia}{m}{y}' '{ia}{kh}'
+      '{o}{v}{i}' '{o}{m}'
+      '{e}{ts}{soft}' '{e}{m}'
+      '{e}{n}{soft}'
+      '{o}{i`}'
+      '{i}{iu}' '{i}{yi}'
+      '{i}{v}'
+      '{e}{v}' '{o}{v}' '{e}{i`}' '{y}{ia}{m}' '{y}{ia}{kh}' '{y}{iu}'
+      '{i}{ia}{m}' '{i}{ia}{kh}' '{i}{ia}'
+      '{y}{ia}' '{ye}{iu}' '{e}{v}{i}' '{ye}{m}' '{e}{yi}{v}' '{yi}{v}'
+      '{i}{ye}{iu}' '{y}{ye}{iu}' '{e}{ye}{iu}'
+      '{k}{a}' '{k}{y}' '{ts}{i}' '{k}{u}' '{k}{o}' '{o}{k}' // female gender
+      '{soft}{yi}' '{soft}{ye}' '{soft}{ye}{iu}' '{soft}{iu}' '{soft}{ia}'
+        (delete)
+    )
+  )
+
+  define remove_last_letter as (
+    [substring] among (
+      '{a}' '{v}' '{e}' '{ye}' '{y}' '{i}' '{yi}' // NOUN: {v}{o}{d}|{a}, VERB: {v}{ch}{y}|{v}
+      '{i`}' '{o}' '{u}' '{soft}' '{iu}' '{ia}'
+        (delete)
+    )
+  )
+
+  define remove_last_2_letters as ( // 2-letters from the all previous sets
+    [substring] among (
+      '{a}{v}' '{a}{m}' '{a}{r}' '{a}{kh}' '{a}{ch}'
+      '{e}{v}' '{e}{ye}' '{e}{i`}' '{e}{m}' '{e}{n}' '{e}{sh}' '{e}{iu}'
+      '{ye}{ye}' '{ye}{m}' '{ye}{sh}' '{ye}{iu}'
+      '{y}{i`}' '{y}{m}' '{y}{kh}' '{y}{sh}' '{y}{iu}' '{y}{ia}'
+      '{i}{v}' '{yi}{v}' '{i}{yi}' '{i}{i`}' '{yi}{i`}' '{i}{m}' '{i}{kh}' '{i}{iu}'
+      '{yi}{sh}'
+      '{l}{a}' '{l}{y}' '{l}{o}' '{l}{iu}' // {a}{k}{u}|{l}{i} ?
+      '{m}{e}' // {d}{i}{ia}|{m}{y} ?
+      '{o}{v}' '{o}{yi}' '{o}{i`}' '{o}{m}' '{o}{iu}' '{s}{soft}' '{s}{ia}'
+       //'{t}{y}' '{t}{e}' '{t}{soft}' // need to remove ?
+      '{u}{ch}'
+      '{soft}{ye}' '{soft}{yi}' '{soft}{iu}' '{soft}{ia}'
+      '{iu}{ch}'
+      '{ia}{m}' '{ia}{kh}' '{ia}{ch}'
+        (delete)
+    )
+  )
+
+  define remove_last_vowel as (
+    [substring] among (
+      '{a}' '{e}' '{ye}' '{y}' '{i}' '{yi}'
+      '{i`}' '{o}' '{u}' '{soft}' '{iu}' '{ia}'
+        (delete)
+    )
+  )
+
+  define remove_last_2_vowels as (
+    remove_vowel_before_vowel
+      and remove_last_vowel
+  )
+
+  define remove_vowel_before_vowel as (
+    [substring] among (
+      '{a}' '{e}' '{ye}' '{y}' '{i}' '{yi}' '{i`}' '{o}' '{u}' '{soft}' '{iu}' '{ia}'
+      ('{a}' or '{e}' or '{ye}' or '{y}' or '{i}' or '{yi}' or '{i`}' or '{o}' or '{u}' or '{soft}' or '{iu}' or '{ia}' delete )
```

Comment:
A long `or` chain is less efficient - better to replace this line with an `among` which can check for a set of n strings in O(log(n)) instead of O(n):

```
      ( among ('{a}' '{e}' '{ye}' '{y}' '{i}' '{yi}' '{i`}' '{o}' '{u}' '{soft}' '{iu}' '{ia}') delete )
```

Reply:
Looking at this again, more efficient still would be to use a grouping.  Above add `vowel` to the `groupings` list, then define it as:

```
define vowel as v + '{i`}{soft}'
```

(Maybe `vowel` is a bad name for this if `v` is the "real" vowels.  Or maybe these two should actually just be in `v` anyway?)

Then this function becomes:

```
define remove_vowel_before_vowel as (
    [vowel] vowel delete
)
```

The other `among` uses where it's just a list of individual characters with a single common action could be done similarly.

The snowball compiler could be smarter and turn such an `among` into a `grouping` but the Snowball code for the grouping version actually seems clearer.

(It looks to me like this function is a bit misnamed as it actually seems to remove a vowel which is *after* a vowel since it's working in `backwardmode`, but if I follow the code it probably would be both clearer and more efficient to eliminate this function and make `remove_last_2_vowels` just do `[vowel vowel] delete`).